src/Aspire.Hosting/build/Aspire.Hosting.targets

```diff
@@ -104,14 +98,83 @@ internal static class ]]>%(ClassName)<![CDATA[
+  <UsingTask TaskName="GetNonExecutableReferences"
+             TaskFactory="RoslynCodeTaskFactory"
+             AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
+    <ParameterGroup>
+      <AppProjectTargetFramework ParameterType="Microsoft.Build.Framework.ITaskItem[]" Required="true" />
+      <NonExecutableReferences ParameterType="Microsoft.Build.Framework.ITaskItem[]" Output="true" />
+    </ParameterGroup>
+    <Task>
+      <Using Namespace="System.Xml.Linq" />
+      <Code Type="Fragment" Language="cs">
+        <![CDATA[
+          HashSet<ITaskItem> nonExecutableReferences = new HashSet<ITaskItem>();
+          
+          foreach (var appProject in AppProjectTargetFramework)
+          {
+            var additionalProperties = appProject.GetMetadata("AdditionalPropertiesFromProject");
+            if (string.IsNullOrEmpty(additionalProperties))
+            {
+              // Skip any projects that don't contain the right metadata
+              continue;
+            }
+                
+            var additionalPropertiesXml = XElement.Parse(additionalProperties);
+            foreach (var targetFrameworkElement in additionalPropertiesXml.Elements())
+            {
+              var isExe = targetFrameworkElement.Element("_IsExecutable");
+              if (isExe != null && !string.Equals(isExe.Value, "true", StringComparison.OrdinalIgnoreCase))
+              {
+                nonExecutableReferences.Add(appProject);
+              }
+            }
+          }
+
+          NonExecutableReferences = nonExecutableReferences.ToArray();
+        ]]>
+      </Code>
+    </Task>
+  </UsingTask>
+
+  <!--
+  Validates that all the ProjectReferences of an Aspire AppHost project are executables and
+  informs the developer to set 'AspireHostLibrary=true' if they really intended on ProjectReferencing a library.
+  -->
+  <Target Name="_ValidateAspireHostServiceReferences"
+          DependsOnTargets="_CreateServiceAppProjects"
+          Condition="'$(IsAspireHost)' == 'true' and '$(SkipValidateAspireHostServiceReferences)' != 'true'">
+
+    <MSBuild Projects="@(_ServiceAppProject)"
+             Targets="GetTargetFrameworks"
+             BuildInParallel="$(BuildInParallel)"
+             Properties="%(_ServiceAppProject.SetConfiguration); %(_ServiceAppProject.SetPlatform)"
+             ContinueOnError="!$(BuildingProject)"
+             RemoveProperties="%(_ServiceAppProject.GlobalPropertiesToRemove);TargetFramework;RuntimeIdentifier;SelfContained;$(_GlobalPropertiesToRemoveFromProjectReferences)"
+             SkipNonexistentTargets="true">
+      <Output TaskParameter="TargetOutputs" ItemName="_ServiceAppProjectTargetFramework" />
+    </MSBuild>
+
+    <GetNonExecutableReferences AppProjectTargetFramework="@(_ServiceAppProjectTargetFramework)">
+      <Output TaskParameter="NonExecutableReferences" ItemName="_NonExecutableServiceApps" />
+    </GetNonExecutableReferences>
+
+    <Warning Code="ASPIRE004"
```

Comment:
Should this be an error instead? Will there ever be the case where someone references a non-exe and it doesn't want to reference the real assembly? I think even for IHostedServices that won't be the case.

Reply:
I think going to an error would be too far. You can shut this behavior off, which typically isn't possible for "errors". It isn't typically an error that we can't continue. It is just something you probably didn't intend.

Reply:
okay, fair enough.